sklearn/utils/_sorting.pxd

```diff
@@ -7,3 +8,25 @@ cdef int simultaneous_sort(
     ITYPE_t *idx,
     ITYPE_t size,
 ) nogil
+
+cdef void sort(cnp.npy_float32* Xf, cnp.npy_intp* samples, cnp.npy_intp n) nogil
+
+cdef void swap(
+    cnp.npy_float32* Xf,
+    cnp.npy_intp* samples,
+    cnp.npy_intp i,
+    cnp.npy_intp j,
+) nogil
+
+cdef cnp.npy_float32 median3(cnp.npy_float32* Xf, cnp.npy_intp n) nogil
+
+cdef void introsort(cnp.npy_float32* Xf, cnp.npy_intp *samples, cnp.npy_intp n, int maxd) nogil
+
+cdef void sift_down(
+    cnp.npy_float32* Xf,
+    cnp.npy_intp* samples,
+    cnp.npy_intp start,
+    cnp.npy_intp end,
+) nogil
```

Comment:
Do we really need to export the low-level helper functions like `sift_down`, `swp`, `median3` in the `.pxd` file?

Also I don't think we need to export both `sort` and `introsort` and I would find it more intuitive to rename:

- `sort` to `introsort` (the main API to call from other modules in scikit-learn),
- `introsort` to `introsort_recursion` or `_introsort` or similar to make it more explicit that this function is not meant to be used directly but only useful to expose the prototype imposed by the recursive calls in the algorithm.

Reply:
By the way I think we could add a few lines of doc for each sorting function (assuming we stop exposing the helpers).

Reply:
I did iterations locally that I have not yet pushed and that are treating most of your comments.